packages/dds/tree/src/simple-tree/treeApi.ts

```diff
@@ -142,11 +176,82 @@ export const treeNodeApi: TreeNodeApi = {
+ * to allow efficiently searching for new content (and initial content) of a given type.
+ *
  * @public
  */
-export interface TreeNodeEvents {
+export interface TreeChangeEvents {
 	/**
-	 * Raised on a node right after a change is applied to one of its fields or the fields of a descendant node.
+	 * Emitted by a node when:
+	 *
+	 * - The content of one or more of the node's properties changes (i.e., the property now contains a new node, or
+	 * nothing if it previously contained a node).
+	 *
+	 * - For an array node, when an element is added, removed, or moved.
+	 *
+	 * - For a map node, when a key is added, updated, or removed.
+	 *
+	 * @remarks
+	 * In particular, this event is not raised when:
+	 *
+	 * - Properties of a child node change. Notably, updates to an array node or a map node (like adding or removing
+	 * elements/entries) will raise this event on the array/map node itself, but not on the node that contains the
+	 * array/map node as one of its properties.
+	 *
+	 * - The node is moved to a different location in the tree or removed from the tree.
+	 * In this case the event is raised on the _parent_ node, not the node itself.
+	 *
+	 * Also note that value nodes do not have properties (child nodes), so this event is never raised for them.
+	 *
+	 * For remote edits, this event is not guaranteed to occur in the same order or quantity that it did in
+	 * the client that made the original edit.
+	 * While a batch of edits will as a whole update the tree to the appropriate end state, no guarantees are made about
+	 * how many times this event will be raised during any intermediate states.
+	 * When it is raised, the tree is guaranteed to be in-schema.
+	 *
+	 * @privateRemarks
+	 * This event occurs whenever the apparent contents of the node instance change, regardless of what caused the change.
+	 * For example, it will fire when the local client reassigns a child, when part of a remote edit is applied to the
+	 * node, or when the node has to be updated due to resolution of a merge conflict
+	 * (for example a previously applied local change might be undone, then reapplied differently or not at all).
+	 */
+	afterShallowChange(): void;
+
+	/**
+	 * Emitted by a node when changes happen anywhere in the subtree rooted at it.
+	 *
+	 * @remarks
+	 * The node itself is part of the subtree, so this event will be emitted even if the only changes are to the properties
+	 * of the node itself.
+	 *
+	 * This event is not raised when the node itself is moved to a different location in the tree or removed from the tree.
+	 * In that case it is raised on the _parent_ node, not the node itself.
+	 *
+	 * It may fire at a time when the change(s) that triggered it are not yet visible if the listener inspects the tree.
+	 * In that case, it is guaranteed to fire again after the change(s) _are_ visible to the listener.
+	 *
```

Comment:
This information might sound kind of scary to a user trying to read this for the first time. It seems to imply that you'll need some kind of complicated event-queueing in order to process the events correctly. When really, we're just trying to say "changes cause events to fire, but events can also fire without a change having happened. **Simply update your lookaside state according to the state of the tree after each event fires.** Don't try to count events or rely on events firing at certain times." 

I wonder if we can communicate this more simply.

Reply:
How about "it may fire multiple times for a given batch of changes, but the changes are not guaranteed to be visible to the listener _every_ time the event fires".

Regarding "can also fire without a change having happened", does a delta visit traverse a node if the value assigned to one of its properties is the same it already has (thus no effective change)? Or when batching several changes, if there is no net effect (e.g. update a number node from 2 to 3 then back to 2... with the same node identity?)? @yann-achard-MS that one might be for you. If we do visit nodes even in those cases, then I can see an argument for "the event can fire when no changes happened" although the "node identity" bit might come into play (the value might be the same, but the actual node object in the tree is now a different one; is that considered "a change"?).

Reply:
We should reserve the right to fire the event even if nothing has changed, now or in the future. Otherwise, it's on us to detect all possible ways that "the end state ends up the same as the beginning state" e.g. of a transaction and in that case, not emit an event. It seems to me we should have the policy of "of course we'll do our best not to fire unnecessary events, but that's a matter of optimization, not a matter of correctness."

There's a key idea here that I'm hoping we can figure out how to communicate. It is derived like this:

1. Ignore events for the moment. Suppose we have no events at all. A user can still use the tree via a "pull" model. They have some external state (e.g. a UI) that is derived from our tree state, and they can refresh ("pull") at some frequency. Each time they refresh/pull, they read the tree state to update their external state. This works okay, but they'll probably refresh/pull more often than they need to, leading to over-inval.
2. So, we introduce events. Our events fire (_at least_) when the tree state changes. That means that the user no longer needs to "pull" and instead can rely on our events doing a "push" - we tell the user when the state changes. The user listens to those events and updates their external state by reading the tree state. This is not that different than the "pull" model, the key is that we have told the user what the frequency is at which they need to update their external state. We push at least as often as we need to _guarantee that they could do no better than if they were manually pulling_. That is, in so many words, the policy.

I am hoping we can communicate that idea somehow (obviously much less verbosely than how I said it above). Right now we're throwing the events at the user and trying to guess what assumptions they are going to make. Like, "they'll probably think the event fires exactly once, so let's tell them that it might fire multiple times." But that could be confusing to a user who _didn't_ assume it would fire exactly once. Or, "if the event fires twice, then they'll assume that both times it's the same and they can get everything they need from the first event, so let's tell them that actually they might need to look at the second event too." But, that is confusing to a user who _didn't_ assume that. Instead, I want us to be able to tell them the true expectations that they should have, but we're struggling to do so concisely. 

"You may have external state, derived from the tree state, that you wish to keep maximally synchronized. In that case, you can subscribe to our events which will guarantee maximal synchronization."

What do you think? Do you agree? How do we express that intuitively on these events? It's not a unique or novel concept - any time that you are synchronizing a UI with a data model you walk this same problem space.

Reply:
In theses cases we do indeed consider it a change because of node identity, so the delta visit will treat these changes the same as any other (meaning the fact that the values are the same is not taken into account).

Reply:
Regarding

> "You may have external state, derived from the tree state, that you wish to keep maximally synchronized. In that case, you can subscribe to our events which will guarantee maximal synchronization."
>
> What do you think? Do you agree? How do we express that intuitively on these events? It's not a unique or novel concept - any time that you are synchronizing a UI with a data model you walk this same problem space.

And

> "We should reserve the right to fire the event even if nothing has changed, now or in the future."

I don't disagree with the assessment of the intention, and probably would not mind as much if the event in question was `potentialSubtreeChange` or `subtreeInvalidation` or something along those lines... but for one called specifically `afterDeepChange` I can't avoid finding it weird that it may fire when nothing has actually happened in the tree yet. Documentation for `subtreeInvalidation` to me would feel reasonable saying "Emitted when something in the subtree _may_ have changed.". Could that be the solution we're looking for? `afterShallowChange` + `subtreeInvalidation`?

Reply:
That's a really good point about the name. Maybe we should have `synchronizeNode` (shallow) and `synchronizeTree` (deep) or something.

Reply:
+1 on Noah's proposal to tell users the true expectations that they should have. I think we can still add a complement of information so long as it's clear that this information can be safely ignored so long as they have the right expectations (e.g., "The event may fire even though no change has been applied to the tree.").

+1 on Alejandro's point on the event name. I don't like `potentialSubtreeChange`, I'm fine with `subtreeInvalidation`. I would like to add `subtreeDirty` to the list of options. I think "Dirty" is well accepted as requiring an update but not implying that the update will make a difference.

Reply:
> "You may have external state, derived from the tree state, that you wish to keep maximally synchronized. In that case, you can subscribe to our events which will guarantee maximal synchronization."

@noencke - what's the "maximally" here? What is being maximized/minimized?

Reply:
@yann-achard-MS Yes, exactly, that's what I meant.